source/validate_atomics.cpp

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates correctness of atomic SPIR-V instructions.
+
+#include "validate.h"
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+// Validates correctness of atomic instructions.
+spv_result_t AtomicsPass(ValidationState_t& _,
+                             const spv_parsed_instruction_t* inst) {
+  const SpvOp opcode = static_cast<SpvOp>(inst->opcode);
+  const uint32_t result_type = inst->type_id;
+
+  switch (opcode) {
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor: {
+      if (opcode == SpvOpAtomicLoad || opcode == SpvOpAtomicExchange ||
+          opcode == SpvOpAtomicCompareExchange) {
+        if (!_.IsFloatScalarType(result_type) &&
+            !_.IsIntScalarType(result_type)) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+              << spvOpcodeString(opcode)
+              << ": expected Result Type to be int or float scalar type";
+        }
+      } else {
+        if (!_.IsIntScalarType(result_type)) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+              << spvOpcodeString(opcode)
+              << ": expected Result Type to be int scalar type";
+        }
+      }
+
+      const uint32_t pointer_type = _.GetOperandTypeId(inst, 2);
+
+      uint32_t data_type = 0;
+      uint32_t storage_class = 0;
+      if (!_.GetPointerTypeInfo(pointer_type, &data_type, &storage_class)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Pointer to be of type OpTypePointer";
+      }
+
+      if (data_type != result_type) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
```

Comment:
Nit: Maybe it's just my eyes. But I generally find it a little bit hard to associate the error message with the the problematic parameter since we don't have clear marking of which parameter is wrong. I'd suggest putting the parameter names in `<>` to make it clear, e.g., `expected <Pointer> to point to a value of type <Result Type>`. What's better, add colors to them! :)

Reply:
I already have hundreds of error messages which use capitalization to mark keywords. I don't want to have to change it everywhere.

Reply:
That's fine.